File: article/scripts/loop.py
#!/usr/bin/env python3
import argparse
import difflib
import json
import subprocess
import sys
import pathlib
import time
from datetime import datetime
from typing import Optional

try:
    import yaml  # pip install pyyaml
except ImportError:
    sys.exit("Please `pip install pyyaml` to use article/scripts/loop.py")

ROOT = pathlib.Path(__file__).resolve().parents[2]
CFG = yaml.safe_load((ROOT / "article/config/loop.yaml").read_text())
PROMPT_VALUES = {
    "PROJECT_NAME": CFG["project"]["name"],
    "RESULTS_DIR": CFG["paths"]["results_dir"],
    "MANUSCRIPT_MAIN": CFG["paths"]["manuscript_main"],
    "MANUSCRIPT_SUP": CFG["paths"]["manuscript_sup"],
    "BIB_PATH": CFG["paths"]["references_bib"],
    "SPG_PATH": CFG["paths"]["spg"],
    "OUTLINE_PATH": CFG["paths"]["outline"],
    "DOCS_DIR": CFG["paths"]["docs_dir"],
    "OUTLINE_STRATEGY": CFG["planner"]["outline_strategy"],
    "IMAGES_DIR": CFG["paths"]["images_dir"],
    "REVIEW_DIR": "article/artifacts",
    "AUTHOR_GUIDELINES": "",
}
GUIDELINES_DIR = CFG["paths"].get("guidelines_dir")
CODEX_CONFIG_PATH = ROOT / ".tooling" / "codex" / "config.toml"


def load_codex_overrides():
    if not CODEX_CONFIG_PATH.exists():
        return []
    overrides = []
    current_section = None
    for raw_line in CODEX_CONFIG_PATH.read_text().splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#"):
            continue
        if line.startswith("[") and line.endswith("]"):
            current_section = line[1:-1].strip()
            continue
        if "=" not in line:
            continue
        key, value = line.split("=", 1)
        full_key = f"{current_section}.{key.strip()}" if current_section else key.strip()
        overrides.extend(["-c", f"{full_key}={value.strip()}"])
    return overrides


CODEX_OVERRIDES = load_codex_overrides()
STEP_INDEX = {"codex": 1, "claude": 2, "gemini": 3}


def run(cmd, cwd=ROOT, interactive=False):
    print(">>", " ".join(cmd), flush=True)
    if interactive:
        r = subprocess.run(cmd, cwd=cwd)
        if r.returncode != 0:
            sys.exit(r.returncode)
        return ""

    r = subprocess.run(cmd, cwd=cwd, text=True, capture_output=True)
    if r.returncode != 0:
        print(r.stdout)
        print(r.stderr, file=sys.stderr)
        sys.exit(r.returncode)
    return r.stdout


def render_template(path):
    text = pathlib.Path(path).read_text()
    for key, value in PROMPT_VALUES.items():
        text = text.replace(f"{{{{{key}}}}}", str(value))
    return text


def load_author_guidelines() -> list[str]:
    if not GUIDELINES_DIR:
        return []
    base = ROOT / GUIDELINES_DIR
    if not base.exists():
        return []
    parts: list[str] = []
    for candidate in sorted(base.rglob("*.md")):
        try:
            text = candidate.read_text().strip()
            if text:
                parts.append(text)
        except OSError:
            continue
    return parts


def latest_review_blob():
    artifacts_dir = ROOT / "article" / "artifacts"
    if not artifacts_dir.exists():
        return None
    candidates = sorted(artifacts_dir.glob("review_*.json"))
    if not candidates:
        return None
    latest = max(candidates, key=lambda p: p.stat().st_mtime)
    try:
        content = latest.read_text().strip()
    except OSError:
        return None
    return latest.name, content


def extract_review_from_stream(stream_text: str) -> Optional[str]:
    """Parse Gemini stream-json output and return the embedded review JSON."""
    if not stream_text:
        return None
    assembled = []
    for line in stream_text.splitlines():
        line = line.strip()
        if not line:
            continue
        # attempt to parse top-level object
        try:
            obj = json.loads(line)
        except json.JSONDecodeError:
            continue
        if obj.get("type") == "message" and obj.get("role") == "assistant":
            content = obj.get("content", "")
            if content:
                assembled.append(content)
    if not assembled:
        return None
    payload = "".join(assembled)
    if "```json" in payload:
        start = payload.index("```json") + len("```json")
        end = payload.find("```", start)
        if end == -1:
            body = payload[start:]
        else:
            body = payload[start:end]
        return body.strip()
    # fall back: payload itself may already be JSON
    payload = payload.strip()
    if not payload:
        return None
    try:
        parsed = json.loads(payload)
        return json.dumps(parsed, indent=2)
    except json.JSONDecodeError:
        return payload


def codex(prompt_path, mode="headless"):
    prompt = render_template(prompt_path)

    if mode == "interactive":
        print(f"[codex] Prompt from {prompt_path}:\n{prompt}\n")
        cmd = ["codex", *CODEX_OVERRIDES, "exec", "--cd", str(ROOT), "--", prompt]
        return run(cmd, interactive=True)

    cmd = [
        "codex",
        *CODEX_OVERRIDES,
        "exec",
        "--cd",
        str(ROOT),
        "--json",
        "--",
        prompt,
    ]
    return run(cmd)


def claude(prompt_path, mode="headless"):
    prompt = render_template(prompt_path)
    review_info = latest_review_blob()
    if review_info:
        filename, blob = review_info
        prompt = (
            f"{prompt}\n\n"
            f"---\nLatest Gemini review (`{filename}`):\n```json\n{blob}\n```\n"
            "Address each point (or log TODOs) before handing off.\n"
        )

    base_args = [
        "--allow-dangerously-skip-permissions",
        "--dangerously-skip-permissions",
        "--permission-mode",
        "acceptEdits",
    ]

    if mode == "interactive":
        print(f"[claude] Prompt from {prompt_path}:\n{prompt}\n")
        cmd = [
            "claude",
            "--verbose",
            *base_args,
            "-p",
            "--output-format",
            "stream-json",
            prompt,
        ]
        print(
            "+ streaming Claude output (ctrl+C to abort; transcript logged above)\n",
            flush=True,
        )
        proc = subprocess.Popen(
            cmd,
            cwd=ROOT,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
        )
        buffer: list[str] = []
        try:
            assert proc.stdout is not None
            for line in proc.stdout:
                print(line, end="", flush=True)
                buffer.append(line)
        finally:
            proc.wait()
        if proc.returncode != 0:
            sys.exit(proc.returncode)
        return "".join(buffer)

    return run(
        ["claude", "--verbose", *base_args, "-p", "--output-format", "json", prompt],
    )


def gemini(prompt_path, mode="headless"):
    prompt = render_template(prompt_path)

    if mode == "interactive":
        print(f"[gemini] Prompt from {prompt_path}:\n{prompt}\n")
        cmd = ["gemini", "--output-format", "stream-json", prompt]
        print(
            "+ streaming Gemini output (ctrl+C to abort; review will be saved when complete)\n",
            flush=True,
        )
        proc = subprocess.Popen(
            cmd,
            cwd=ROOT / "article",
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
        )
        buffer = []
        try:
            assert proc.stdout is not None
            for line in proc.stdout:
                print(line, end="", flush=True)
                buffer.append(line)
        finally:
            proc.wait()
        stream_text = "".join(buffer)
        if proc.returncode != 0:
            sys.exit(proc.returncode)
        extracted = extract_review_from_stream(stream_text)
        return extracted if extracted else stream_text

    cmd = ["gemini", "--output-format", "json", prompt]
    return run(cmd, cwd=ROOT / "article")


def build_pdf():
    try:
        run(["latexmk", "-pdf", "-quiet", CFG["paths"]["manuscript_main"]], cwd=ROOT)
    except FileNotFoundError:
        print("[loop] latexmk not available; skipping PDF build.", file=sys.stderr)


def main():
    parser = argparse.ArgumentParser(description="Codex → Claude → Gemini loop orchestrator")
    parser.add_argument(
        "--mode",
        choices=["headless", "interactive"],
        default="headless",
        help="Run automation in headless mode (default) or launch interactive CLIs.",
    )
    parser.add_argument(
        "--start-from",
        choices=["codex", "claude", "gemini"],
        default="codex",
        help="Skip agents before the selected stage.",
    )
    parser.add_argument(
        "--build-pdf",
        action="store_true",
        help="Run latexmk to compile the manuscript after Claude edits.",
    )
    parser.add_argument(
        "--guideline",
        action="append",
        default=[],
        help="Inline author guideline (can be provided multiple times).",
    )
    parser.add_argument(
        "--guideline-file",
        action="append",
        default=[],
        help="Path to a text/markdown file containing author guidelines (repeatable).",
    )
    args = parser.parse_args()
    mode = args.mode
    initial_start_index = STEP_INDEX[args.start_from]

    iters = int(CFG["loop"]["max_iters"])
    gates = CFG.get("readability_gates", {})
    clarity_min = float(gates.get("clarity_min", 0))
    storytelling_min = float(gates.get("storytelling_min", 0))
    readability_min = float(gates.get("readability_min", 0))
    max_conf = int(gates.get("max_confusions", sys.maxsize))
    max_style = int(gates.get("max_style_findings", sys.maxsize))

    guideline_chunks: list[str] = []
    for inline in args.guideline:
        text = inline.strip()
        if text:
            guideline_chunks.append(text)
    for file_path in args.guideline_file:
        try:
            text = (ROOT / file_path).read_text().strip() if not pathlib.Path(file_path).is_absolute() else pathlib.Path(file_path).read_text().strip()
            if text:
                guideline_chunks.append(text)
        except OSError as exc:
            print(f"[loop] warning: failed to read guideline file '{file_path}': {exc}", file=sys.stderr)
    guideline_chunks.extend(load_author_guidelines())
    combined_guidelines = "\n\n".join(guideline_chunks)
    if not combined_guidelines.strip():
        combined_guidelines = "(no explicit author guidelines provided)"
    PROMPT_VALUES["AUTHOR_GUIDELINES"] = combined_guidelines

    manuscript_relpaths = [
        pathlib.Path(PROMPT_VALUES["MANUSCRIPT_MAIN"]),
        pathlib.Path(PROMPT_VALUES["MANUSCRIPT_SUP"]),
        pathlib.Path(PROMPT_VALUES["BIB_PATH"]),
    ]
    manuscript_paths = [ROOT / rel for rel in manuscript_relpaths]
    original_manuscript: dict[pathlib.Path, Optional[str]] = {}
    for abs_path in manuscript_paths:
        try:
            original_manuscript[abs_path] = abs_path.read_text()
        except OSError:
            original_manuscript[abs_path] = None

    timings = {
        "codex_graph": 0.0,
        "codex_outline": 0.0,
        "claude": 0.0,
        "build_pdf": 0.0,
        "gemini": 0.0,
    }
    overall_start = time.time()

    for i in range(1, iters + 1):
        print(f"\n==== LOOP {i} ({mode}) ====\n")
        loop_start_index = initial_start_index if i == 1 else STEP_INDEX["codex"]

        if loop_start_index <= STEP_INDEX["codex"]:
            t0 = time.time()
            codex("article/scripts/prompts/codex_graph.txt", mode=mode)
            timings["codex_graph"] += time.time() - t0
            t0 = time.time()
            codex("article/scripts/prompts/codex_linearize.txt", mode=mode)
            timings["codex_outline"] += time.time() - t0
        else:
            print("[loop] Skipping Codex stages (start-from>=claude).")

        if loop_start_index <= STEP_INDEX["claude"]:
            t0 = time.time()
            claude("article/scripts/prompts/claude_draft.txt", mode=mode)
            timings["claude"] += time.time() - t0
        else:
            print("[loop] Skipping Claude stage (start-from=gemini).")

        if args.build_pdf:
            t0 = time.time()
            build_pdf()
            timings["build_pdf"] += time.time() - t0

        t0 = time.time()
        review_json = gemini("article/scripts/prompts/gemini_blind_review.txt", mode=mode)
        timings["gemini"] += time.time() - t0

        try:
            review = json.loads(review_json)
        except Exception:
            review = json.loads(review_json.strip().splitlines()[-1])

        artifacts_dir = ROOT / "article" / "artifacts"
        artifacts_dir.mkdir(parents=True, exist_ok=True)
        timestamp = datetime.utcnow().strftime("%Y%m%dT%H%M%SZ")
        review_path = artifacts_dir / f"review_loop{i}_{timestamp}.json"
        review_path.write_text(json.dumps(review, indent=2))

        confs = len(review.get("confusions", []))
        clarity = review.get("clarity_score", 0)
        storytelling = review.get("storytelling_score", 0)
        readability = review.get("readability_score", 0)
        style_findings = review.get("style_findings", [])
        style_count = len(style_findings) if isinstance(style_findings, list) else 0
        if (
            CFG["loop"]["stop_on_success"]
            and clarity >= clarity_min
            and storytelling >= storytelling_min
            and readability >= readability_min
            and confs <= max_conf
            and style_count <= max_style
        ):
            print(
                "[loop] stop: "
                f"clarity={clarity} storytelling={storytelling} "
                f"readability={readability} confusions={confs} "
                f"style_findings={style_count}"
            )
            break
    else:
        print("[loop] finished without meeting gates")

    print("\n==== Manuscript Diff Summary ====")
    for rel_path, abs_path in zip(manuscript_relpaths, manuscript_paths):
        before = original_manuscript.get(abs_path)
        try:
            after = abs_path.read_text()
        except OSError:
            after = None

        display_name = str(rel_path)
        if before is None and after is None:
            print(f"- {display_name}: unavailable")
            continue
        if before is None:
            added_lines = len(after.splitlines()) if after is not None else 0
            print(f"- {display_name}: created ({added_lines} lines)")
            continue
        if after is None:
            print(f"- {display_name}: deleted")
            continue
        if before == after:
            print(f"- {display_name}: no textual changes")
            continue

        diff_lines = list(
            difflib.unified_diff(
                before.splitlines(),
                after.splitlines(),
                fromfile=f"before/{display_name}",
                tofile=f"after/{display_name}",
                lineterm="",
            )
        )
        max_lines = 120
        truncated = len(diff_lines) > max_lines
        snippet = diff_lines[:max_lines]
        print(f"- {display_name} diff:")
        for line in snippet:
            print(f"  {line}")
        if truncated:
            print(f"  ... ({len(diff_lines) - max_lines} more diff lines)")

    total_elapsed = time.time() - overall_start
    print("\n==== Timing Summary ====")
    print(f"Total elapsed: {total_elapsed:.1f}s")
    for key, secs in timings.items():
        print(f"- {key}: {secs:.1f}s")


if __name__ == "__main__":
    main()

File: loop-runner/loop.ts
#!/usr/bin/env node
import fs from "fs";
import path from "path";
import { spawn } from "child_process";
import { fileURLToPath } from "url";
import yaml from "js-yaml";
import { query } from "@anthropic-ai/claude-agent-sdk";

type ClaudeAgentOptions = Record<string, any>;
type Message = Record<string, any>;

const __filename = fileURLToPath(import.meta.url);
const RUNNER_DIR = path.dirname(__filename);
const ROOT = path.resolve(RUNNER_DIR, "..");
const CONFIG_PATH = path.join(RUNNER_DIR, "config", "loop.yaml");
const CFG = yaml.load(fs.readFileSync(CONFIG_PATH, "utf8")) as any;

const PROMPT_VALUES: Record<string, string> = {
  PROJECT_NAME: CFG.project.name,
  RESULTS_DIR: CFG.paths.results_dir,
  MANUSCRIPT_MAIN: CFG.paths.manuscript_main,
  MANUSCRIPT_SUP: CFG.paths.manuscript_sup,
  BIB_PATH: CFG.paths.references_bib,
  SPG_PATH: CFG.paths.spg,
  OUTLINE_PATH: CFG.paths.outline,
  DOCS_DIR: CFG.paths.docs_dir,
  OUTLINE_STRATEGY: CFG.planner.outline_strategy,
  IMAGES_DIR: CFG.paths.images_dir,
  REVIEW_DIR: "article/artifacts",
  AUTHOR_GUIDELINES: "",
  ARXIV_DIR: CFG.paths.arxiv_dir ?? "article/arxiv/papers",
  LITERATURE_INDEX: CFG.paths.literature_index ?? "article/arxiv/index.json",
};

const STEP_INDEX: Record<string, number> = {
  research: 1,
  codex: 2,
  claude: 3,
  gemini_audit: 4,
  gemini: 5,
};

type RunResult = { output: string; code: number };

function renderTemplate(filePath: string): string {
  let text = fs.readFileSync(filePath, "utf8");
  Object.entries(PROMPT_VALUES).forEach(([key, value]) => {
    text = text.replace(new RegExp(`{{${key}}}`, "g"), String(value));
  });
  return text;
}

function loadAuthorGuidelines(): string[] {
  const dir = CFG.paths.guidelines_dir;
  if (!dir) return [];
  const abs = path.join(ROOT, dir);
  if (!fs.existsSync(abs)) return [];
  const files = fs.readdirSync(abs).filter((f) => f.endsWith(".md"));
  const chunks: string[] = [];
  for (const file of files.sort()) {
    try {
      const txt = fs.readFileSync(path.join(abs, file), "utf8").trim();
      if (txt) chunks.push(txt);
    } catch {
      continue;
    }
  }
  return chunks;
}

function latestBlob(prefix: string): [string, string] | null {
  const dir = path.join(ROOT, "article", "artifacts");
  if (!fs.existsSync(dir)) return null;
  const candidates = fs
    .readdirSync(dir)
    .filter((f) => f.startsWith(prefix) && f.endsWith(".json"))
    .map((f) => path.join(dir, f));
  if (!candidates.length) return null;
  const latest = candidates.sort(
    (a, b) => fs.statSync(b).mtimeMs - fs.statSync(a).mtimeMs,
  )[0];
  try {
    return [path.basename(latest), fs.readFileSync(latest, "utf8").trim()];
  } catch {
    return null;
  }
}

const latestReviewBlob = () => latestBlob("review_");
const latestCitationAuditBlob = () => latestBlob("citation_audit_");

function handleJsonLine(line: string): boolean {
  try {
    const obj = JSON.parse(line);
    const msg = obj?.message;
    const content = msg?.content ?? obj?.content;
    const parts: string[] = [];
    if (Array.isArray(content)) {
      for (const p of content) {
        if (p?.type === "text" && p.text) parts.push(p.text);
      }
    } else if (typeof content === "string" && content.trim()) {
      parts.push(content.trim());
    }
    if (parts.length) {
      console.log(parts.join("\n"));
      return true;
    }
  } catch {
    return false;
  }
  return false;
}

function runCommand(cmd: string[], opts: { cwd?: string } = {}): Promise<RunResult> {
  return new Promise((resolve, reject) => {
    const proc = spawn(cmd[0], cmd.slice(1), {
      cwd: opts.cwd ?? ROOT,
      stdio: ["ignore", "pipe", "pipe"],
    });
    let output = "";
    let buffer = "";
    const handleChunk = (chunk: Buffer) => {
      const str = chunk.toString();
      output += str;
      buffer += str;
      const lines = buffer.split(/\r?\n/);
      buffer = lines.pop() || "";
      for (const line of lines) {
        if (!line.trim()) continue;
        if (!handleJsonLine(line)) {
          console.log(line);
        }
      }
    };
    proc.stdout.on("data", handleChunk);
    proc.stderr.on("data", handleChunk);
    proc.on("close", (code) => {
      if (buffer.trim()) {
        if (!handleJsonLine(buffer)) console.log(buffer.trim());
        output += buffer;
      }
      resolve({ output, code: code ?? 0 });
    });
    proc.on("error", reject);
  });
}

function extractReviewFromStream(streamText: string): string | null {
  if (!streamText) return null;
  const assembled: string[] = [];
  for (const line of streamText.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    try {
      const obj = JSON.parse(trimmed);
      if (obj?.type === "message" && obj?.role === "assistant") {
        const content = obj.content;
        if (typeof content === "string" && content) assembled.push(content);
        if (Array.isArray(content)) {
          for (const part of content) {
            if (part?.type === "text" && part.text) assembled.push(part.text);
          }
        }
      }
    } catch {
      continue;
    }
  }
  if (!assembled.length) return null;
  const payload = assembled.join("");
  if (payload.includes("```json")) {
    const start = payload.indexOf("```json") + "```json".length;
    const end = payload.indexOf("```", start);
    const body = payload.slice(start, end === -1 ? undefined : end).trim();
    return body;
  }
  return payload.trim() || null;
}

function parseJsonPayload(payload: string): any {
  const cleaned = (payload || "").trim();
  if (!cleaned) throw new Error("empty payload");
  try {
    return JSON.parse(cleaned);
  } catch {
    const lines = cleaned.split(/\r?\n/);
    return JSON.parse(lines[lines.length - 1]);
  }
}

async function runCodex(promptPath: string) {
  const prompt = renderTemplate(promptPath);
  await runCommand([
    "codex",
    "exec",
    "--cd",
    ROOT,
    "--json",
    "--",
    prompt,
  ]);
}

async function runClaudeDraft() {
  let prompt = renderTemplate(path.join(RUNNER_DIR, "prompts", "claude_draft.txt"));
  const review = latestReviewBlob();
  const audit = latestCitationAuditBlob();
  if (review) {
    prompt += `\n\n---\nLatest Gemini review (\`${review[0]}\`):\n\`\`\`json\n${review[1]}\n\`\`\`\nAddress each point (or log TODOs) before handing off.\n`;
  }
  if (audit) {
    prompt += `\n\n---\nLatest citation audit (\`${audit[0]}\`):\n\`\`\`json\n${audit[1]}\n\`\`\`\nResolve misplacements/missing citations where possible; otherwise leave explicit TODOs.\n`;
  }
  await runCommand([
    "claude",
    "--verbose",
    "--allow-dangerously-skip-permissions",
    "--dangerously-skip-permissions",
    "--permission-mode",
    "acceptEdits",
    "-p",
    "--output-format",
    "stream-json",
    prompt,
  ]);
}

async function runGemini(promptPath: string, extraContext = ""): Promise<string> {
  const base = renderTemplate(promptPath);
  const prompt = extraContext ? `${base}\n\n${extraContext}` : base;
  const { output } = await runCommand([
    "gemini",
    "--output-format",
    "stream-json",
    prompt,
  ], { cwd: path.join(ROOT, "article") });
  return output;
}

function buildResearchOptions(): ClaudeAgentOptions {
  let mcpServers: Record<string, any> = {};
  const mcpPath = path.join(ROOT, ".mcp.json");
  if (fs.existsSync(mcpPath)) {
    try {
      const raw = JSON.parse(fs.readFileSync(mcpPath, "utf8"));
      if (raw?.mcpServers && typeof raw.mcpServers === "object") {
        mcpServers = raw.mcpServers;
      }
    } catch {
      mcpServers = {};
    }
  }
  const debugDir = process.env.CLAUDE_DEBUG_DIR || path.join(ROOT, ".claude", "debug");
  try {
    fs.mkdirSync(debugDir, { recursive: true });
  } catch {
    // ignore
  }
  return {
    model: "sonnet-4-5",
    permissionMode: "acceptEdits",
    allowedTools: [
      "Read(article/backbone/**)",
      "Read(article/manuscript/**)",
      "Read(article/docs/**)",
      "Read(article/arxiv/**)",
      "Write(article/arxiv/index.json)",
      "Write(article/arxiv/papers/**)",
      "WebSearch",
      "WebFetch",
      "mcp:arxiv/*",
    ],
    disallowedTools: [
      "Edit(**/*)",
      "Bash(**/*)",
      "NotebookEdit(**/*)",
      "Write(**/*)",
    ],
    cwd: ROOT,
    mcpServers,
    settingSources: ["project"],
    env: {
      ...process.env,
      CLAUDE_DEBUG_DIR: debugDir,
    },
    pathToClaudeCodeExecutable: process.env.CLAUDE_PATH || "claude",
    stderr: (data: string) => {
      const trimmed = data.trim();
      if (trimmed) console.error(trimmed);
    },
  };
}

async function runClaudeResearch(): Promise<any> {
  const promptPath = path.join(RUNNER_DIR, "prompts", "claude_research.txt");
  const prompt = renderTemplate(promptPath);
  const opts = buildResearchOptions();
  const messages: Message[] = [];
  for await (const msg of query({ prompt, options: opts })) {
    messages.push(msg as Message);
    const content = (msg as any).content;
    if (Array.isArray(content)) {
      const texts = content
        .filter((p) => p?.type === "text" && p.text)
        .map((p) => p.text);
      if (texts.length) console.log(texts.join("\n"));
    } else if (typeof content === "string" && content.trim()) {
      console.log(content.trim());
    }
  }
  for (let i = messages.length - 1; i >= 0; i--) {
    const content = (messages[i] as any).content;
    if (Array.isArray(content)) {
      for (const block of content) {
        if (block?.type === "text" && block.text) {
          try {
            const parsed = JSON.parse(block.text);
            return parsed;
          } catch {
            continue;
          }
        }
      }
    } else if (typeof content === "string" && content.trim()) {
      try {
        return JSON.parse(content.trim());
      } catch {
        continue;
      }
    }
  }
  return { session_id: "", fetched: [], skipped: [], todos: [] };
}

function applyAuthorGuidelines(extra: string[]) {
  const combined = extra.filter(Boolean).join("\n\n");
  PROMPT_VALUES.AUTHOR_GUIDELINES = combined || "(no explicit author guidelines provided)";
}

function readInlineArgs(): { startFrom: string; buildPdf: boolean; guidelines: string[]; guidelineFiles: string[] } {
  const args = process.argv.slice(2);
  const guidelines: string[] = [];
  const guidelineFiles: string[] = [];
  let startFrom = "research";
  let buildPdf = false;
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === "--start-from" && args[i + 1]) {
      startFrom = args[++i];
    } else if (arg === "--build-pdf") {
      buildPdf = true;
    } else if (arg === "--guideline" && args[i + 1]) {
      guidelines.push(args[++i]);
    } else if (arg === "--guideline-file" && args[i + 1]) {
      guidelineFiles.push(args[++i]);
    }
  }
  return { startFrom, buildPdf, guidelines, guidelineFiles };
}

function readGuidelineFiles(files: string[]): string[] {
  const chunks: string[] = [];
  for (const file of files) {
    const abs = path.isAbsolute(file) ? file : path.join(ROOT, file);
    try {
      const txt = fs.readFileSync(abs, "utf8").trim();
      if (txt) chunks.push(txt);
    } catch {
      continue;
    }
  }
  return chunks;
}

async function main() {
  const { startFrom, buildPdf, guidelines, guidelineFiles } = readInlineArgs();
  const startIndex = STEP_INDEX[startFrom] ?? STEP_INDEX.research;

  const guidelineChunks = [
    ...guidelines.map((g) => g.trim()).filter(Boolean),
    ...readGuidelineFiles(guidelineFiles),
    ...loadAuthorGuidelines(),
  ];
  applyAuthorGuidelines(guidelineChunks);

  const manuscripts = [
    path.join(ROOT, PROMPT_VALUES.MANUSCRIPT_MAIN),
    path.join(ROOT, PROMPT_VALUES.MANUSCRIPT_SUP),
    path.join(ROOT, PROMPT_VALUES.BIB_PATH),
  ];
  const original: Record<string, string | null> = {};
  for (const m of manuscripts) {
    try {
      original[m] = fs.readFileSync(m, "utf8");
    } catch {
      original[m] = null;
    }
  }

  const timings: Record<string, number> = {
    research: 0,
    codex_graph: 0,
    codex_outline: 0,
    claude: 0,
    build_pdf: 0,
    gemini_audit: 0,
    gemini: 0,
  };
  const overallStart = Date.now();
  const artifactsDir = path.join(ROOT, "article", "artifacts");
  fs.mkdirSync(artifactsDir, { recursive: true });

  const gates = CFG.readability_gates ?? {};
  const clarityMin = Number(gates.clarity_min ?? 0);
  const storytellingMin = Number(gates.storytelling_min ?? 0);
  const readabilityMin = Number(gates.readability_min ?? 0);
  const maxConf = Number(gates.max_confusions ?? Number.MAX_SAFE_INTEGER);
  const maxStyle = Number(gates.max_style_findings ?? Number.MAX_SAFE_INTEGER);

  const iters = Number(CFG.loop.max_iters ?? 1);

  for (let i = 1; i <= iters; i++) {
    console.log(`\n==== LOOP ${i} ====\n`);
    const loopStart = i === 1 ? startIndex : STEP_INDEX.research;
    let auditPayload: string | null = null;

    if (loopStart <= STEP_INDEX.research) {
      const t0 = Date.now();
      console.log("[loop] Running Claude Research (SDK)");
      const researchJson = await runClaudeResearch();
      const ts = new Date().toISOString().replace(/[-:]/g, "").replace(/\.\d+Z$/, "Z");
      const p = path.join(artifactsDir, `research_loop${i}_${ts}.json`);
      fs.writeFileSync(p, JSON.stringify(researchJson, null, 2));
      timings.research += (Date.now() - t0) / 1000;
    } else {
      console.log("[loop] Skipping Claude research stage (start-from>=codex).");
    }

    if (loopStart <= STEP_INDEX.codex) {
      const t0 = Date.now();
      console.log("[loop] Running Codex Graph");
      await runCodex(path.join(RUNNER_DIR, "prompts", "codex_graph.txt"));
      timings.codex_graph += (Date.now() - t0) / 1000;
      const t1 = Date.now();
      console.log("[loop] Running Codex Outline");
      await runCodex(path.join(RUNNER_DIR, "prompts", "codex_linearize.txt"));
      timings.codex_outline += (Date.now() - t1) / 1000;
    } else {
      console.log("[loop] Skipping Codex stages (start-from>=claude).");
    }

    if (loopStart <= STEP_INDEX.claude) {
      const t0 = Date.now();
      console.log("[loop] Running Claude Draft");
      await runClaudeDraft();
      timings.claude += (Date.now() - t0) / 1000;
    } else {
      console.log("[loop] Skipping Claude stage (start-from>=gemini_audit).");
    }

    if (buildPdf) {
      const t0 = Date.now();
      console.log("[loop] Building PDF");
      await runCommand(["latexmk", "-pdf", "-quiet", PROMPT_VALUES.MANUSCRIPT_MAIN], { cwd: ROOT });
      timings.build_pdf += (Date.now() - t0) / 1000;
    }

    if (loopStart <= STEP_INDEX.gemini_audit) {
      const t0 = Date.now();
      console.log("[loop] Running Gemini Citation Audit");
      const auditRaw = await runGemini(path.join(RUNNER_DIR, "prompts", "gemini_citation_audit.txt"));
      const extracted = extractReviewFromStream(auditRaw);
      const audit = parseJsonPayload(extracted ?? auditRaw);
      auditPayload = JSON.stringify(audit, null, 2);
      const ts = new Date().toISOString().replace(/[-:]/g, "").replace(/\.\d+Z$/, "Z");
      fs.writeFileSync(path.join(artifactsDir, `citation_audit_loop${i}_${ts}.json`), auditPayload);
      timings.gemini_audit += (Date.now() - t0) / 1000;
    } else {
      console.log("[loop] Skipping Gemini citation audit (start-from=gemini).");
      const latest = latestCitationAuditBlob();
      if (latest) auditPayload = latest[1];
    }

    const auditContext = auditPayload
      ? `---\nCitation audit findings preceding this review:\n\`\`\`json\n${auditPayload}\n\`\`\`\nIncorporate these findings; ensure misplacements/missing citations are reflected in your feedback.\n`
      : "";

    const t0 = Date.now();
    console.log("[loop] Running Gemini Blind Review");
    const reviewRaw = await runGemini(
      path.join(RUNNER_DIR, "prompts", "gemini_blind_review.txt"),
      auditContext,
    );
    const extractedReview = extractReviewFromStream(reviewRaw);
    const review = parseJsonPayload(extractedReview ?? reviewRaw);
    const ts = new Date().toISOString().replace(/[-:]/g, "").replace(/\.\d+Z$/, "Z");
    fs.writeFileSync(path.join(artifactsDir, `review_loop${i}_${ts}.json`), JSON.stringify(review, null, 2));
    timings.gemini += (Date.now() - t0) / 1000;

    const confs = (review.confusions ?? []).length;
    const clarity = Number(review.clarity_score ?? 0);
    const storytelling = Number(review.storytelling_score ?? 0);
    const readability = Number(review.readability_score ?? 0);
    const styleCount = Array.isArray(review.style_findings) ? review.style_findings.length : 0;
    if (
      CFG.loop.stop_on_success &&
      clarity >= clarityMin &&
      storytelling >= storytellingMin &&
      readability >= readabilityMin &&
      confs <= maxConf &&
      styleCount <= maxStyle
    ) {
      console.log(
        `[loop] stop: clarity=${clarity} storytelling=${storytelling} readability=${readability} confusions=${confs} style_findings=${styleCount}`,
      );
      break;
    }
  }

  console.log("\n==== Manuscript Diff Summary ====");
  for (const m of manuscripts) {
    const rel = path.relative(ROOT, m);
    const before = original[m];
    let after: string | null = null;
    try {
      after = fs.readFileSync(m, "utf8");
    } catch {
      after = null;
    }
    if (before == null && after == null) {
      console.log(`- ${rel}: unavailable`);
      continue;
    }
    if (before == null) {
      const added = after ? after.split(/\r?\n/).length : 0;
      console.log(`- ${rel}: created (${added} lines)`);
      continue;
    }
    if (after == null) {
      console.log(`- ${rel}: deleted`);
      continue;
    }
    if (before === after) {
      console.log(`- ${rel}: no textual changes`);
      continue;
    }
    console.log(`- ${rel}: changed (diff not shown)`);
  }

  const totalElapsed = (Date.now() - overallStart) / 1000;
  console.log("\n==== Timing Summary ====");
  console.log(`Total elapsed: ${totalElapsed.toFixed(1)}s`);
  Object.entries(timings).forEach(([k, v]) => console.log(`- ${k}: ${v.toFixed(1)}s`));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
